reco_utils/recommender/rbm/rbm.py

```diff
@@ -0,0 +1,946 @@
+
+        #--------------Sampling protocol for Gibbs sampling-----------------------------------
+        self.k=1 #initialize the G_sampling step
+        self.l=0 #initialize epoch_sample index
+        #-------------------------Main algo---------------------------
+
+        self.G_sampling() #returns the sampled value of the visible units
+
+        obj = self.Losses(self.v) #objective function
+        rate = self.alpha/self.minibatch  #learning rate rescaled by the batch size
+
+        #Instantiate the momentum optimizer
+        opt = tf.contrib.optimizer_v2.AdamOptimizer(learning_rate = rate).minimize(loss= obj)
+
+        pvh, vp = self.infere() #sample the value of the visible units given the hidden. Also returns  the related probabilities
+
+        if self.with_metrics_:#if true (default) returns evaluation metrics
+            Mse_train = [] #Lists to collect the metrics across epochs
+            #Metrics
+            Mserr  = self.msr_error(self.v_k)
+            Clacc  = self.accuracy(self.v_k)
+
+        if self.save_model_: #save the model to file
+            saver = tf.train.Saver()
+
+        init_g = tf.global_variables_initializer() #Initialize all variables in the graph
+
+        #Config GPU memory
+        Config = tf.ConfigProto(log_device_placement = True, allow_soft_placement=True)
+        Config.gpu_options.per_process_gpu_memory_fraction = 0.5
+
+        #Start TF training session on default graph
+        self.sess =  tf.Session(config = Config)
+        self.sess.run(init_g)
+
+        self.sess.run(self.iter.initializer, feed_dict={self.vu: xtr, self.batch_size_: self.minibatch})
+
+        if self.with_metrics_: #this condition is for benchmarking, remove for production
+
+            #start loop over training epochs
+            for i in range(self.epochs):
+
+                epoch_tr_err =0 #initialize the training error for each epoch to zero
+
+                self.Gibbs_protocol(i) #updates the number of sampling steps in Gibbs sampling
+
+                for l in range(num_minibatches):
+
+                    _, batch_err = self.sess.run([opt, Mserr])
+
+                    epoch_tr_err += batch_err/num_minibatches #average mse error per minibatch
+
+                if i % self.display ==0:
+                    print('training epoch %i rmse Train %f' %(i, epoch_tr_err) )
+
+                #write metrics across epochs
+                Mse_train.append(epoch_tr_err) # mse training error per training epoch
+
+            #Evaluates precision on the train and test set
+            precision_train = self.sess.run(Clacc)
+
+            self.sess.run(self.iter.initializer, feed_dict={self.vu: xtst, self.batch_size_: xtst.shape[0]})
+            precision_test = self.sess.run(Clacc)
+            rmse_test = self.sess.run(Mserr)
+
+            elapsed = self.time()
+
+            log.info("done training, Training time %f2" %elapsed)
+
+            #Display training error as a function of epochs
```

Comment:
I think we should extract the plots from the fit, in case the user doesn't want to plot anything

Reply:
yes, the plot and metrics are already optional. By default they are not evaluated/shown unless one uses `with_metrics = True` in the model initialization. Without metrics the training is slighlty faster; In the notebooks I explain that this parameter is optional and I recommend to print metrics to inspect learning.